File: env/engine.py
"""Core simulation engine: the weekly physics of the beer distribution game.

See CLAUDE.md for the authoritative six-phase weekly sequence. Echelons are
indexed 0=Retailer, 1=Wholesaler, 2=Distributor, 3=Factory; each echelon's
supplier is index+1 and its customer is index-1 (or the external customer
for the Retailer, and no supplier for the Factory).
"""

from collections import deque
from dataclasses import dataclass, field
from typing import Callable, Deque, List

import pandas as pd

from agents.interface import Agent, Observation
from env.config import ECHELON_NAMES, SimulationConfig


@dataclass
class EchelonState:
    """Mutable per-echelon state.

    Not every field is used by every role: the Retailer never uses
    incoming_orders (customer demand arrives directly, with no transit
    delay) and the Factory never uses incoming_shipments (it has no
    supplier; incoming_production is its arriving-stock pipeline instead).
    """

    inventory: int
    backlog: int = 0
    supply_line: int = 0
    incoming_orders: Deque[int] = field(default_factory=deque)
    incoming_shipments: Deque[int] = field(default_factory=deque)
    incoming_production: Deque[int] = field(default_factory=deque)


class BeerGameEngine:
    """Runs the four-echelon serial supply chain for a configured horizon."""

    def __init__(
        self,
        config: SimulationConfig,
        agents: List[Agent],
        demand_fn: Callable[[int], int],
    ) -> None:
        if len(agents) != config.n_echelons:
            raise ValueError(f"expected {config.n_echelons} agents, got {len(agents)}")
        self.config = config
        self.agents = agents
        self.demand_fn = demand_fn
        self.factory_index = config.n_echelons - 1
        self.states: List[EchelonState] = [
            EchelonState(
                inventory=config.initial_inventory,
                incoming_orders=(
                    deque([0] * config.order_delay, maxlen=config.order_delay)
                    if i != 0
                    else deque()
                ),
                incoming_shipments=(
                    deque([0] * config.ship_delay, maxlen=config.ship_delay)
                    if i != self.factory_index
                    else deque()
                ),
                incoming_production=(
                    deque([0] * config.production_delay, maxlen=config.production_delay)
                    if i == self.factory_index
                    else deque()
                ),
            )
            for i in range(config.n_echelons)
        ]
        # The two system-boundary counters the conservation test relies on:
        # production is the only source of new units, customer delivery is
        # the only sink. Both start at zero; the fixed initial_inventory
        # loaded above is accounted for separately by the test.
        self.cumulative_produced = 0
        self.cumulative_delivered_to_customer = 0

    def step(self, week: int) -> List[dict]:
        """Advance the simulation by one week; returns one log row per echelon.

        All echelons read pipeline/demand state as it stood at the start of
        the week (phases 1-5); every pipeline push (phase 6) is applied only
        after all echelons have computed their shipment and order for the
        week, so per-echelon iteration order cannot affect the result.
        """
        n = self.config.n_echelons
        states = self.states

        # Phases 1-2: receive arriving shipment/production, then the
        # downstream order (or, for the Retailer, exogenous customer
        # demand), adding it to backlog.
        received_stock = [0] * n
        received_order = [0] * n
        for i, st in enumerate(states):
            if i == self.factory_index:
                qty = st.incoming_production.popleft()
            else:
                qty = st.incoming_shipments.popleft()
            st.inventory += qty
            st.supply_line -= qty
            received_stock[i] = qty

            order_in = self.demand_fn(week) if i == 0 else st.incoming_orders.popleft()
            st.backlog += order_in
            received_order[i] = order_in

        # Phase 3: ship to satisfy backlog as far as inventory allows.
        shipped = [0] * n
        for i, st in enumerate(states):
            qty = min(st.inventory, st.backlog)
            st.inventory -= qty
            st.backlog -= qty
            shipped[i] = qty

        # Phase 4: record cost.
        cost = [
            self.config.holding_cost * st.inventory + self.config.backlog_cost * st.backlog
            for st in states
        ]

        # Phase 5: each agent decides its new order from the post-shipment
        # state, using only its own observation (never another echelon's
        # state).
        order_placed = [0] * n
        for i, st in enumerate(states):
            obs = Observation(
                echelon=ECHELON_NAMES[i],
                week=week,
                inventory=st.inventory,
                backlog=st.backlog,
                supply_line=st.supply_line,
                incoming_order=received_order[i],
            )
            order_placed[i] = self.agents[i].order(obs)

        # Phase 6: propagate this week's shipments and orders, and update
        # supply lines and the two boundary counters. Applied only now, so
        # nothing pushed here was visible to any agent's observation above.
        for i, st in enumerate(states):
            st.supply_line += order_placed[i]

            if i == 0:
                self.cumulative_delivered_to_customer += shipped[i]
            else:
                states[i - 1].incoming_shipments.append(shipped[i])

            if i == self.factory_index:
                st.incoming_production.append(order_placed[i])
                self.cumulative_produced += order_placed[i]
            else:
                states[i + 1].incoming_orders.append(order_placed[i])

        return [
            {
                "week": week,
                "echelon": ECHELON_NAMES[i],
                "inventory": states[i].inventory,
                "backlog": states[i].backlog,
                "supply_line": states[i].supply_line,
                "shipment_received": received_stock[i],
                "order_received": received_order[i],
                "shipped": shipped[i],
                "order_placed": order_placed[i],
                "cost": cost[i],
                "cumulative_produced": self.cumulative_produced,
                "cumulative_delivered_to_customer": self.cumulative_delivered_to_customer,
            }
            for i in range(n)
        ]

    def run(self) -> pd.DataFrame:
        """Run the full configured horizon and return a tidy log DataFrame."""
        rows = []
        for week in range(1, self.config.horizon_weeks + 1):
            rows.extend(self.step(week))
        return pd.DataFrame(rows)

File: js/engine.js
// Core simulation engine: the weekly physics of the beer distribution
// game. Ported faithfully from env/engine.py -- same six-phase weekly
// sequence, same supply-line accounting (its own accumulator, incremented
// on order and decremented on receipt, never derived from the pipeline
// contents), same delay pipelines (plain arrays used as FIFOs: push() to
// append, shift() for popleft()). Echelons are indexed
// 0=Retailer, 1=Wholesaler, 2=Distributor, 3=Factory; each echelon's
// supplier is index+1 and its customer is index-1 (or the external
// customer for the Retailer, and no supplier for the Factory).
//
// Output row field names are intentionally snake_case (order_placed, not
// orderPlaced) to match the Python engine's logged column names exactly --
// this is the ground truth format tests/js_reference/reference_run.json
// uses, and tests/js_reference/verify.mjs diffs against it field-by-field.

import { ECHELON_NAMES } from "./config.js";

class EchelonState {
  constructor({ inventory, incomingOrders, incomingShipments, incomingProduction }) {
    this.inventory = inventory;
    this.backlog = 0;
    this.supplyLine = 0;
    this.incomingOrders = incomingOrders;
    this.incomingShipments = incomingShipments;
    this.incomingProduction = incomingProduction;
  }
}

export class BeerGameEngine {
  constructor(config, agents, demandFn) {
    if (agents.length !== config.nEchelons) {
      throw new Error(`expected ${config.nEchelons} agents, got ${agents.length}`);
    }
    this.config = config;
    this.agents = agents;
    this.demandFn = demandFn;
    this.factoryIndex = config.nEchelons - 1;

    this.states = [];
    for (let i = 0; i < config.nEchelons; i++) {
      this.states.push(
        new EchelonState({
          inventory: config.initialInventory,
          incomingOrders: i !== 0 ? new Array(config.orderDelay).fill(0) : [],
          incomingShipments: i !== this.factoryIndex ? new Array(config.shipDelay).fill(0) : [],
          incomingProduction: i === this.factoryIndex ? new Array(config.productionDelay).fill(0) : [],
        })
      );
    }

    // The two system-boundary counters the conservation test relies on:
    // production is the only source of new units, customer delivery is
    // the only sink.
    this.cumulativeProduced = 0;
    this.cumulativeDeliveredToCustomer = 0;
  }

  // Phases 1-4: receive arriving shipment/production, receive the
  // downstream order (or, for the Retailer, exogenous customer demand),
  // ship to satisfy backlog as far as inventory allows, and record cost.
  // None of this depends on any echelon's phase-5 decision, so it can
  // always run to completion regardless of where each order will come
  // from -- this is what lets a turn-based caller (js/gameController.js)
  // pause for a human's decision between this and finishWeek() without
  // touching the physics at all.
  beginWeek(week) {
    const n = this.config.nEchelons;
    const states = this.states;

    const receivedStock = new Array(n).fill(0);
    const receivedOrder = new Array(n).fill(0);
    for (let i = 0; i < n; i++) {
      const st = states[i];
      const qty = i === this.factoryIndex ? st.incomingProduction.shift() : st.incomingShipments.shift();
      st.inventory += qty;
      st.supplyLine -= qty;
      receivedStock[i] = qty;

      const orderIn = i === 0 ? this.demandFn(week) : st.incomingOrders.shift();
      st.backlog += orderIn;
      receivedOrder[i] = orderIn;
    }

    const shipped = new Array(n).fill(0);
    for (let i = 0; i < n; i++) {
      const st = states[i];
      const qty = Math.min(st.inventory, st.backlog);
      st.inventory -= qty;
      st.backlog -= qty;
      shipped[i] = qty;
    }

    const cost = states.map((st) => this.config.holdingCost * st.inventory + this.config.backlogCost * st.backlog);

    // The observation every agent (algorithmic or human) receives for its
    // phase-5 decision -- deliberately just these fields, nothing about
    // any other echelon. shipmentReceived and cost are additive relative
    // to the original engine (existing agents ignore both; only the game
    // controller's restricted player view reads them).
    const observations = states.map((st, i) => ({
      echelon: ECHELON_NAMES[i],
      week,
      inventory: st.inventory,
      backlog: st.backlog,
      supplyLine: st.supplyLine,
      incomingOrder: receivedOrder[i],
      shipmentReceived: receivedStock[i],
      cost: cost[i],
    }));

    return { receivedStock, receivedOrder, shipped, cost, observations };
  }

  // Phase 6: propagate this week's shipments and orders, update supply
  // lines and the two boundary counters, and return the logged rows.
  // orderPlaced must already be a complete array (one entry per echelon)
  // -- where each entry came from (an algorithmic agent, or a human via
  // the game controller) is irrelevant here.
  finishWeek(week, weekState, orderPlaced) {
    const n = this.config.nEchelons;
    const states = this.states;
    const { receivedStock, receivedOrder, shipped, cost } = weekState;

    for (let i = 0; i < n; i++) {
      const st = states[i];
      st.supplyLine += orderPlaced[i];

      if (i === 0) {
        this.cumulativeDeliveredToCustomer += shipped[i];
      } else {
        states[i - 1].incomingShipments.push(shipped[i]);
      }

      if (i === this.factoryIndex) {
        st.incomingProduction.push(orderPlaced[i]);
        this.cumulativeProduced += orderPlaced[i];
      } else {
        states[i + 1].incomingOrders.push(orderPlaced[i]);
      }
    }

    return states.map((st, i) => ({
      week,
      echelon: ECHELON_NAMES[i],
      inventory: st.inventory,
      backlog: st.backlog,
      supply_line: st.supplyLine,
      shipment_received: receivedStock[i],
      order_received: receivedOrder[i],
      shipped: shipped[i],
      order_placed: orderPlaced[i],
      cost: cost[i],
      cumulative_produced: this.cumulativeProduced,
      cumulative_delivered_to_customer: this.cumulativeDeliveredToCustomer,
    }));
  }

  step(week) {
    const weekState = this.beginWeek(week);
    const orderPlaced = weekState.observations.map((obs, i) => this.agents[i].order(obs));
    return this.finishWeek(week, weekState, orderPlaced);
  }

  run() {
    const rows = [];
    for (let week = 1; week <= this.config.horizonWeeks; week++) {
      rows.push(...this.step(week));
    }
    return rows;
  }
}
